### PR TITLE
WIP: Custom container

### PR DESCRIPTION
**Pull request by kravcik**
- [ ]  Add test
- [ ] Change to `customForm` (Form responsibility create custom container)?

Do you think this need deeper look? I don't think function `addCustomContainer($form, $name)` is necessary - better option would be add customForm and implement this function there.

This can be temporary solution.

### CODE REVIEW

File: src/DI/MultiplierExtension.php [hunk 1]
```diff
@@ -11,7 +11,8 @@ class MultiplierExtension extends CompilerExtension {
 
 	/** @var array */
 	public $defaults = [
-		'name' => 'addMultiplier'
+		'name' => 'addMultiplier',
+		'customContainer' => 'Nette\Forms\Container'
```
Comment (MartkCz) [inline src/DI/MultiplierExtension.php]: `=> Container::class` imo is better

File: src/Multiplier.php [hunk 2]
```diff
@@ -71,6 +71,9 @@ class Multiplier extends Container {
 	/** @var Container[] */
 	protected $noValidate = [];
 
+	/** @var Container */
+	public static $container = 'Nette\Forms\Container';
```
Comment (MartkCz) [inline src/Multiplier.php]: same here `= Container::class`

File: src/Multiplier.php [hunk 3]
```diff
@@ -565,8 +582,9 @@ public function setValues($values, $erase = false) {
 	/**
 	 * @param string $name
 	 */
-	public static function register($name = 'addMultiplier') {
-		Container::extensionMethod($name, function (Container $form, $name, $factory, $copyNumber = 1, $maxCopies = null) {
+	public static function register($name = 'addMultiplier', $container = 'Nette\Forms\Container') {
```
Comment (MartkCz) [inline src/Multiplier.php]: same here

**Comment by MartkCz** [comment]
Change to customForm (Form responsibility create custom container)?

But how? the only solution that comes to mind: create interface IFormContainerProvider, form implements the interface and multiplier checks if form implement interface...

File: src/DI/MultiplierExtension.php (hunk 1 above)
Comment (kravcik) [reply src/DI/MultiplierExtension.php]: Good point

**Comment by kravcik** [comment]
I just have a "strange" feeling about function (`addCustomContainer()`). Thats all. Origin idea was something like this.

```
protected function applyDefaultValues(Container $container) {
		$form = new CustomForm('_foo_multiplier');
		$factoryContainer = $form->addContainer('foo'); //return customContainer
```

Btw atm it's works well with custom Form and container.
```
$form = new customForm;
$form->addMultiplier('m', function($container, $form)
{
  $form->addWhisperer();
});

```

Should I prepare test for this solution?

**Comment by MartkCz** [comment]
how about

```php
trait TFormComponents {
    public function addContainer(...) {
    // returns CustomContainer
    }

    public function addMultiplier(...) {
    // returns CustomMultiplier
    }

}
```
```php
class CustomMultiplier extends Multiplier {
    use TFormComponents;
}
```

```php
class CustomForm extends Form {
    use TFormComponents;
}
```

```php
class CustomContainer extends Container {
    use TFormComponents;
}
```

**Comment by kravcik** [comment]
Good idea. So I guess I will this implement outside multiplier. Should I close this?

**Comment by MartkCz** [comment]
Thank you for your time :)

**Pull request closed without merging**